psl-core/src/main/java/org/linqs/psl/util/VizDataCollection.java

```diff
@@ -0,0 +1,215 @@
+package org.linqs.psl.util;
+
+import org.linqs.psl.grounding.GroundRuleStore;
+import org.linqs.psl.model.atom.Atom;
+import org.linqs.psl.model.atom.GroundAtom;
+import org.linqs.psl.model.rule.GroundRule;
+import org.linqs.psl.model.rule.Rule;
+import org.linqs.psl.model.rule.UnweightedGroundRule;
+import org.linqs.psl.model.rule.WeightedGroundRule;
+import org.linqs.psl.model.rule.AbstractRule;
+import org.linqs.psl.model.term.Constant;
+import org.linqs.psl.model.term.Variable;
+
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.FilterOutputStream;
+import java.io.PrintStream;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.zip.GZIPOutputStream;
+
+public class VizDataCollection {
+    private static final Logger log = LoggerFactory.getLogger(VizDataCollection.class);
+
+    private static Runtime runtime = null;
+    private static VisualizationData vizData = null;
+
+    private static String outputPath = null;
+
+    static {
+        init();
+    }
+
+    private VizDataCollection() {}
+
+    private static synchronized void init() {
+        if (runtime != null) {
+            return;
+        }
+        vizData = new VisualizationData();
+        runtime = Runtime.getRuntime();
+        runtime.addShutdownHook(new ShutdownHook());
+    }
+
+    public static void outputJSON() throws IOException {
+        FilterOutputStream stream = System.out;
+
+        if (outputPath != null) {
+            try {
+                stream = new GZIPOutputStream(new PrintStream(outputPath));
+            } catch (IOException ex) {
+                throw new RuntimeException(ex);
+            }
+        }
+
+        writeToStream(stream);
+
+        if (outputPath != null) {
+            stream.close();
+        }
+    }
+
+    /**
+     * Write to stream with JSON formatting.
+     */
+    private static void writeToStream(FilterOutputStream stream) throws IOException {
+        // JSON format reference: https://www.json.org/json-en.html.
+        stream.write("{ \"truthMap\" :".getBytes());
+
+        // Write each map as a JSON object, each JSON object is comma delimited.
+        writeMap(stream, vizData.truthMap, "truthMap");
+        stream.write(", \"rules\" :".getBytes());
+        writeMap(stream, vizData.rules, "rules");
+        stream.write(", \"groundRules\" :".getBytes());
+        writeMap(stream, vizData.groundRules, "groundRules");
+        stream.write(", \"groundAtoms\" :".getBytes());
+        writeMap(stream, vizData.groundAtoms, "groundAtoms");
+
+        stream.write('}');
+    }
+
+    /**
+     * Write map to stream with JSON formatting.
```

Comment:
Remember to include information about assumptions.
In this method you are making an important assumption about the nested elements of this map.

Including information about that assumption in the body of the method is fine,
but remember that most people will never look inside a method.
For example, how many times have you looked at the javadocs vs the actual code for the java utils?